mesos-master-detector/src/main/scala/com/mesosphere/mesos/MasterDetector.scala

```diff
@@ -0,0 +1,136 @@
+package com.mesosphere.mesos
+
+import java.net.URL
+import java.nio.charset.StandardCharsets
+import java.util.concurrent.CompletionStage
+
+import com.mesosphere.usi.metrics.Metrics
+import com.mesosphere.usi.storage.zookeeper.PersistenceStore.Node
+import com.mesosphere.usi.storage.zookeeper.{
+  AsyncCuratorBuilderFactory,
+  AsyncCuratorBuilderSettings,
+  ZooKeeperPersistenceStore
+}
+import com.typesafe.scalalogging.StrictLogging
+import org.apache.curator.framework.CuratorFrameworkFactory
+import org.apache.curator.retry.RetryOneTime
+import org.apache.mesos.v1.Protos
+import play.api.libs.functional.syntax._
+import play.api.libs.json._
+import play.api.libs.json.Reads._
+
+import scala.async.Async.{async, await}
+import scala.compat.java8.FutureConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+trait MasterDetector {
+
+  /** @return the Mesos master URL for the cluster. */
+  def getMaster()(implicit ex: ExecutionContext): CompletionStage[URL]
+
+  /** @return whether the master string is valid. */
+  def isValid(): Boolean
+}
+
+object MasterDetector {
+
+  /**
+    * Constructs a master detector base on the master string passed.
+    *
+    * @param master The master string should be one of:
+    *               host:port
+    *               http://host:port
+    *               zk://host1:port1,host2:port2,.../path
+    *               zk://username:password@host1:port1,host2:port2,.../path
+    * @return A master detector.
+    */
+  def apply(master: String, metrics: Metrics): MasterDetector = {
+    if (master.startsWith("zk://")) {
+      Zookeeper(master, metrics)
+    } else {
+      Standalone(master)
+    }
+  }
+}
+
+case class Zookeeper(master: String, metrics: Metrics) extends MasterDetector with StrictLogging {
+  require(master.startsWith("zk://"), s"$master does not start with zk://")
+
+  case class ZkUrl(auth: Option[String], servers: String, path: String)
+
+  implicit val mesosInfoRead: Reads[Protos.MasterInfo] = (
+    (JsPath \ "hostname").read[String] ~
+      (JsPath \ "port").read[Int] ~
+      (JsPath \ "id").read[String] ~
+      (JsPath \ "ip").read[Int]
+  ) { (hostname, port, id, ip) =>
+    Protos.MasterInfo.newBuilder().setHostname(hostname).setPort(port).setId(id).setIp(ip).build()
+  }
+
+  override def isValid(): Boolean = Try(parse()).map(_ => true).getOrElse(false)
+
+  override def getMaster()(implicit ex: ExecutionContext): CompletionStage[URL] = {
+    val ZkUrl(auth, servers, path) = parse()
+
+    val client = {
```

Comment:
`client` should be injected here. The overall project will initialize it and pass into different modules like persistency, mesos-client, (master-detection, leader-election etc.)

Reply:
We cannot inject the client because the credentials come form the connection string. It's a chicken an egg problem.

Reply:
I could inject a factory method: `clientFactory: (authDigest) -> CuratorFramework`.